pkg/kube/snapshot/snapshot_alpha.go

```diff
@@ -0,0 +1,317 @@
+}
+
+// Create creates a VolumeSnapshot and returns it or any error happened meanwhile.
+func (sna *SnapshotAlpha) Create(ctx context.Context, name, namespace, volumeName string, snapshotClass *string, waitForReady bool) error {
+	if _, err := sna.kubeCli.CoreV1().PersistentVolumeClaims(namespace).Get(volumeName, metav1.GetOptions{}); err != nil {
+		if k8errors.IsNotFound(err) {
+			return errors.Errorf("Failed to find PVC %s, Namespace %s", volumeName, namespace)
+		}
+		return errors.Errorf("Failed to query PVC %s, Namespace %s: %v", volumeName, namespace, err)
+	}
+
+	err := sna.createVolumeSnapshot(name, namespace, corev1.ObjectReference{Kind: pvcKind, Name: volumeName}, *snapshotClass)
+	if err != nil {
+		return err
+	}
+
+	if !waitForReady {
+		return nil
+	}
+
+	err = sna.WaitOnReadyToUse(ctx, name, namespace)
+	if err != nil {
+		return err
+	}
+
+	_, err = sna.Get(ctx, name, namespace)
+	return err
+}
+
+// Get will return the VolumeSnapshot in the namespace 'namespace' with given 'name'.
+func (sna *SnapshotAlpha) Get(ctx context.Context, name, namespace string) (*v1alpha1.VolumeSnapshot, error) {
+	us, err := sna.dynCli.Resource(VolSnapGVR).Namespace(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	vs := &v1alpha1.VolumeSnapshot{}
+	err = TransformUnstructured(us, vs)
+	return vs, err
+}
+
+// Delete will delete the VolumeSnapshot and returns any error as a result.
+func (sna *SnapshotAlpha) Delete(ctx context.Context, name, namespace string) error {
+	if err := sna.dynCli.Resource(VolSnapGVR).Namespace(namespace).Delete(name, &metav1.DeleteOptions{}); !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// If the Snapshot does not exist, that's an acceptable error and we ignore it
+	return nil
+}
+
+// Clone will clone the VolumeSnapshot to namespace 'cloneNamespace'.
+// Underlying VolumeSnapshotContent will be cloned with a different name.
+func (sna *SnapshotAlpha) Clone(ctx context.Context, name, namespace, cloneName, cloneNamespace string, waitForReady bool) error {
+	snap, err := sna.Get(ctx, name, namespace)
+	if err != nil {
+		return err
+	}
+	if !snap.Status.ReadyToUse {
+		return errors.Errorf("Original snapshot is not ready, VolumeSnapshot: %s, Namespace: %s", cloneName, cloneNamespace)
+	}
+	if snap.Spec.SnapshotContentName == "" {
+		return errors.Errorf("Original snapshot does not have content, VolumeSnapshot: %s, Namespace: %s", cloneName, cloneNamespace)
+	}
+
+	_, err = sna.Get(ctx, cloneName, cloneNamespace)
+	if err == nil {
+		return errors.Errorf("Target snapshot already exists in target namespace, Volumesnapshot: %s, Namespace: %s", cloneName, cloneNamespace)
+	}
+	if !k8errors.IsNotFound(err) {
+		return errors.Errorf("Failed to query target Volumesnapshot: %s, Namespace: %s: %v", cloneName, cloneNamespace, err)
+	}
+
+	src, err := sna.GetSource(ctx, name, namespace)
```

Comment:
You already fetched the original snapshot once above, this would make a second call to get the same object. I suggest calling sna.getContent. or creating something like get source from snapshot

Reply:
Again you are doing two snapshot.Get() calls. you already have the snapshot object.

Reply:
Tell me why you want to resolve this. I think you can change the signature of GetSource to take a snapshot object and build a source. The only place this is used in K10 is in phase/snapshot.go line 316. and we're calling snapshot get before it there too. Theres no reason to make 2 calls to kubernetes to get the same snapshot

Reply:
I am sorry, I guess clicked "resolve" accidentally. I have another PR which already takes care of the optimisations. But we can address this here itself. In this PR, I have just moved the methods from snapshot.go to snapshot_alpha.go and changed the snapcli to dyncli without making any changes in existing workflow. I was thinking of handling optimisations in another PR